### DISCUSSION TITLE
Read a SMT-LIB2 file and solve it for all  solutions  in C++

### DISCUSSION THREAD

**Discussion by lenchil** [cvc5/cvc5#10454] [Q&A] [answered]
As the title suggests, I want to read in an SMT-LIB2 file and extract all possible solutions. Any suggestions on how to achieve this are welcome. Specifically, I'm struggling with how to retrieve the models from the solver in the code. Since I read in the SMT-LIB file, I don't have an array of sorts and terms readily available. In Z3 this is somewhat easyer becouse of the context and model object

**Comment by ajreynol** [comment]
Unfortunately there isn't a very convenient way to do this at the moment.

Based on your request (and other similar requests we've received), we are planning to add further extensions to our parser API to allow one to retrieve the current set of declared terms and sorts.  This should be available on main in the next few weeks.

I will keep you posted.

**Comment by ajreynol** [comment, marked as answer]
After https://github.com/cvc5/cvc5/commit/2923715aa623234cac90eb529d7084ea1e48b818, we now support `getDeclaredTerms` and `getDeclaredSorts` in the SymbolManager API. This means that after a `checkSat`, you can query what the declared terms that were parsed and subsequently e.g. ask for their values in a model.

Does this suffice for your use case?

**Comment by lenchil** [comment]
Thank you yes this  allowed me to  speedup a implementation by a lot :).
since i now can extract terms and assert blockingclauses at runtime without
having to read in  a modified smt file.

Op di 19 mrt 2024 om 20:12 schreef Andrew Reynolds ***@***.***
